### ISSUE TITLE
Bring Python torch.slice bindings back or add step argument to torch.narrow

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#41625]
Maybe a separate issue is a better place for discussion: https://github.com/pytorch/pytorch/pull/7924#issuecomment-660249077

The usecase: packbits https://github.com/pytorch/pytorch/issues/32867

https://gist.github.com/vadimkantorov/30ea6d278bc492abf6ad328c6965613a#file-packbits-py-L6

**Comment by t-vi** [comment]
Yeah well, I guess everyone is tired from hearing me say things about consistency between C++ and TorchScript by now, so I think I'll just shut up.

**Comment by vadimkantorov** [comment]
At least do not count me in that number :) I'm also in favor of consistency.

**Comment by gchanan** [comment]
don't count me either :).

**Comment by vadimkantorov** [comment]
Currently this:
```python
def tensor_dim_slice(tensor, dim, dim_slice):
    return tensor[(dim if dim >= 0 else dim + tensor.dim()) * (slice(None), ) + (dim_slice, )]
```
fails torch.jit.script when called from a scripted function:

```
    fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(obj))
  File "/miniconda/lib/python3.7/site-packages/torch/jit/_recursive.py", line 568, in try_compile_fn
    return torch.jit.script(fn, _rcb=rcb)
  File "/miniconda/lib/python3.7/site-packages/torch/jit/__init__.py", line 1290, in script
    fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(obj))
RuntimeError:

slice(Tensor 0) -> (Tensor 0):
Expected a value of type 'Tensor' for argument '0' but instead found type 'None'.
:
  File "/root/convasr/ctc.py", line 45
def tensor_dim_slice(tensor, dim, dim_slice):
        return tensor[(dim if dim >= 0 else dim + tensor.dim()) * (slice(None), ) + (dim_slice, )]
                                                            ~~~~~ <--- HERE
```

**Comment by vadimkantorov** [comment]
(created a separate issue #43125 for the JIT issue)

**Comment by vadimkantorov** [comment]
Yep, seems aten::slice exists, but isn't exposed...

**Comment by vadimkantorov** [comment]
Not very related but still: https://github.com/pytorch/pytorch/issues/104314

**Comment by vadimkantorov** [comment]
Another instance is that we currently don't have a nice way of executing `tensor[y : y + h, x : x + w] = 1` in one go as a function. Maybe would be nice to have also a proper exposed alternative to `__setitem__` when needed.

Especially for the advanced indexing for which there appears no alternative in public API:

```python
u = torch.arange(10)
v = reversed(torch.arange(10))
tensor[u, v] = 1
```

**Comment by vadimkantorov** [comment]
Maybe for filling rectangular patch, slice_scatter could be used, if an inplace version existed (https://pytorch.org/docs/master/generated/torch.slice_scatter.html?highlight=slice_scatter#torch.slice_scatter)

Otherwise, maybe needed `slice_fill_`. Or a more general `__setitem__` exposed, indeed

**Comment by vadimkantorov** [comment]
I'm reading some TorchScript code printout.
It has things like:
``` 
_2 = torch.slice(torch.slice(x), 1, None, cutoff)
real_part = torch.to(torch.slice(_2, 2), 6)
_3 = torch.slice(torch.slice(x), 1, cutoff)
imag_part = torch.to(torch.slice(_3, 2), 6)
_4 = torch.add(torch.pow(real_part, 2), torch.pow(imag_part, 2))
...
x3 = torch.slice(torch.slice(x1), 1, None, None, step)
```

what are torch.slice(x), torch.slice(x, d1), torch.slice(x, d1, d2, d3), torch.slice(x, d1, d2, d3, d4) supposed to do?

I also noticed that in TorchScript `.code` printouts there are some methods bound onto `torch.*` namespace which do not exist in normal eager PyTorch. So for trying to somehow to execute/debug these printouts in regular Python code I had to patch it like this:
```
torch.dim = lambda tensor: torch.tensor(tensor.ndim)
torch.__contains__ = lambda ls, x: torch.tensor(torch.tensor(x) in torch.tensor(ls))
torch.__not__ = lambda flag: torch.logical_not(flag)
torch.size = lambda tensor, dim = None: tensor.size(dim) if dim is not None else tensor.size()
torch.view = lambda tensor, *args: tensor.view(*args)
torch.pad = lambda tensor, *args: torch.nn.functional.pad(tensor, *args)
```

Not sure if it's a useful piece of info, but maybe more functional forms (in addition to existing tensor-only method signatures might be helpful in some circumstances, also given that NumPy often has only numpy.method form without any tensor-form)

**Comment by vadimkantorov** [comment]
@BoyuanFeng maybe you would be figuring out soon in the context of your TS migration work what torch.slice was supposed to do in TorchScript :)

**Comment by vadimkantorov** [comment]
It seems that torch.slice is not a thing yet, although `aten.slice` op exists:
- https://github.com/pytorch/pytorch/issues/178618

Although `torch.slice_scatter` seems a thing:
- https://github.com/pytorch/pytorch/issues/177372

`narrow_scatter` might exist, but not `narrow_fill`?
